packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+import type { ReactNode } from 'react';
+
+const history = createBrowserHistory();
+interface Route {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+}
 
-const RoutesContext = createContext< Location | null >( null );
-const HistoryContext = createContext< EnhancedHistory >( history );
+type LocationWithQuery = Location & {
+	query?: Record< string, any >;
+};
+
+interface Match {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+	query?: Record< string, any >;
+	params?: Record< string, any >;
+}
+
+export type BeforeNavigate = ( arg: {
+	path: string;
+	query: Record< string, any >;
+} ) => {
+	path: string;
+	query: Record< string, any >;
+};
+
+interface Config {
+	pathArg: string;
+	beforeNavigate?: BeforeNavigate;
+}
+
+export interface NavigationOptions {
+	transition?: string;
+	state?: Record< string, any >;
+}
+
+const RoutesContext = createContext< Match | null >( null );
+export const ConfigContext = createContext< Config >( { pathArg: 'p' } );
+
+const locationMemo = new WeakMap();
+function getLocationWithQuery() {
+	const location = history.location;
+	let locationWithQuery = locationMemo.get( location );
+	if ( ! locationWithQuery ) {
+		locationWithQuery = {
+			...location,
+			query: Object.fromEntries( new URLSearchParams( location.search ) ),
+		};
+		locationMemo.set( location, locationWithQuery );
+	}
+	return locationWithQuery;
+}
 
 export function useLocation() {
 	return useContext( RoutesContext );
 }
 
 export function useHistory() {
-	return useContext( HistoryContext );
+	const { pathArg, beforeNavigate } = useContext( ConfigContext );
+	return useMemo(
+		() => ( {
+			navigate( rawPath: string, options: NavigationOptions = {} ) {
+				const query = getQueryArgs( rawPath );
+				const path = getPath( 'http://domain.com/' + rawPath ) ?? '';
+				const performPush = () => {
+					const result = beforeNavigate
+						? beforeNavigate( { path, query } )
+						: { path, query };
+					return history.push(
+						{
+							search: buildQueryString( {
+								[ pathArg ]: result.path,
+								...result.query,
+							} ),
+						},
+						options.state
+					);
+				};
+
+				/*
+				 * Skip transition in mobile, otherwise it crashes the browser.
+				 * See: https://github.com/WordPress/gutenberg/pull/63002.
+				 */
+				const isMediumOrBigger =
+					window.matchMedia( '(min-width: 782px)' ).matches;
+				if (
+					! isMediumOrBigger ||
+					// @ts-expect-error
+					! document.startViewTransition ||
+					! options.transition
+				) {
+					return performPush();
+				}
+				document.documentElement.classList.add( options.transition );
+				// @ts-expect-error
+				const transition = document.startViewTransition( () =>
+					performPush()
+				);
+				transition.finished.finally( () => {
+					document.documentElement.classList.remove(
+						options.transition ?? ''
+					);
+				} );
```

Comment:
Could the transition code be extracted into some form of plugin middleware or a filter like `beforeNavigate`? It doesn't look very generic, and could help us design a better extension API for the router. The code can be part of the default configuration, but written as a plugin and overridable.

Reply:
How is it not generic? For me it's actually very generic. You want to apply a "transition" when navigating, you pass it as an arg to `navigate`. It can be extracted of course but for me it's something intrinsic to routing.

Reply:
Agreed on both.

 - It is indeed something that can be plugged, but just because we can, doesn't mean we should. I like to start small and just bundle this personally.
 - I'll update `navigate` to always return a promise.